cpp/src/neighbors/ivf_flat/ivf_flat_build.cuh

```diff
@@ -135,7 +136,10 @@ RAFT_KERNEL build_index_kernel(const LabelT* labels,
   auto source_ix = source_ixs == nullptr ? i + batch_offset : source_ixs[i];
   // In the context of refinement, some indices may be invalid (the generating NN algorithm does
   // not return enough valid items). Do not add the item to the index in this case.
-  if (source_ix == ivf::kInvalidRecord<IdxT> || source_ix == raft::upper_bound<IdxT>()) { return; }
+  if (source_ix == ivf::kInvalidRecord<SourceIdxT> ||
+      source_ix == raft::upper_bound<SourceIdxT>()) {
+    return;
+  }
```

Comment:
I'm a little bit confused whether we should `SourceIdxT` or `IdxT` in here when the types are different (so the `kInvalidRecord` is also different for the two types).
A quick search through the code tells that we only set `kInvalidRecord` in one place: [when we initialize empty IVF lists](https://github.com/rapidsai/cuvs/blob/1155a3a427cd1d1bfaf8fe74a937ed6dfa797ec7/cpp/src/neighbors/ivf_list.cuh#L70-L73). Could you confirm that we indeed initialize the lists in `SourceIdxT` and not in `IdxT`?

Reply:
We had a previous ANN search that gave us the candidates in `source_ixs` in `SourceIdxT` type (i.e we used an `ivf_pq<SourceIdxT>` index). I expect that that the invalid values are marked as `upper_bound<SourceIdxT>()`.